File: react_native/src/services/tmdbService.ts
import { 
  MoviesResponse, 
  DiscoverMovieParams, 
  GenresResponse, 
  MovieDetails, 
  SearchMovieParams 
} from '../types/movie';

const API_BASE_URL = 'https://api.themoviedb.org/3';
const API_TOKEN = 'eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJiZmUwMWE1Y2ZmYzBmNTY3MDZhYmM1NmU4YzNlNzA4ZCIsIm5iZiI6MTc1ODMzNDY5Mi4xNjQsInN1YiI6IjY4Y2UwZWU0MWVjNDQzYjEwMGNkNmU1ZSIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.3Q1ybsIUmSvSikNED11dtMIQ-beSVXfJOMZE-jO8HKg';

class TMDBService {
  private readonly headers = {
    accept: 'application/json',
    Authorization: `Bearer ${API_TOKEN}`,
  };

  private readonly defaultLanguage = 'pt-BR';
  private readonly defaultRegion = 'BR';

  private buildQueryString(params: Record<string, any>): string {
    const searchParams = new URLSearchParams();
    
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        searchParams.append(key, value.toString());
      }
    });
    
    return searchParams.toString();
  }

  private async makeRequest<T>(endpoint: string, params?: Record<string, any>): Promise<T> {
    const defaultParams = {
      language: this.defaultLanguage,
      ...params,
    };

    const queryString = this.buildQueryString(defaultParams);
    const url = `${API_BASE_URL}${endpoint}${queryString ? `?${queryString}` : ''}`;
    
    console.log('🔗 API Request:', url);
    
    try {
      const response = await fetch(url, {
        method: 'GET',
        headers: this.headers,
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      console.log('✅ API Response received');
      return data;
    } catch (error) {
      console.error('❌ API Error:', error);
      throw error;
    }
  }

  /**
   * Get the list of official genres for movies
   */
  async getGenres(): Promise<GenresResponse> {
    return this.makeRequest<GenresResponse>('/genre/movie/list');
  }

  /**
   * Discover movies using over 30 filters and sort options
   */
  async discoverMovies(params: DiscoverMovieParams = {}): Promise<MoviesResponse> {
    const defaultParams = {
      include_adult: false,
      include_video: false,
      page: 1,
      sort_by: 'popularity.desc' as const,
      region: this.defaultRegion,
    };

    const finalParams = { ...defaultParams, ...params };
    return this.makeRequest<MoviesResponse>('/discover/movie', finalParams);
  }

  /**
   * Get a list of movies that are currently in theatres
   */
  async getNowPlayingMovies(page: number = 1): Promise<MoviesResponse> {
    return this.makeRequest<MoviesResponse>('/movie/now_playing', {
      page,
      region: this.defaultRegion,
    });
  }

  /**
   * Get a list of movies ordered by popularity
   */
  async getPopularMovies(page: number = 1): Promise<MoviesResponse> {
    return this.makeRequest<MoviesResponse>('/movie/popular', {
      page,
      region: this.defaultRegion,
    });
  }

  /**
   * Get a list of movies ordered by rating
   */
  async getTopRatedMovies(page: number = 1): Promise<MoviesResponse> {
    return this.makeRequest<MoviesResponse>('/movie/top_rated', {
      page,
      region: this.defaultRegion,
    });
  }

  /**
   * Get a list of movies that are being released soon
   */
  async getUpcomingMovies(page: number = 1): Promise<MoviesResponse> {
    return this.makeRequest<MoviesResponse>('/movie/upcoming', {
      page,
      region: this.defaultRegion,
    });
  }

  /**
   * Get the top level details of a movie by ID
   */
  async getMovieDetails(movieId: number, appendToResponse?: string): Promise<MovieDetails> {
    const params: Record<string, any> = {};
    if (appendToResponse) {
      params.append_to_response = appendToResponse;
    }
    
    return this.makeRequest<MovieDetails>(`/movie/${movieId}`, params);
  }

  /**
   * Search for movies by their original, translated and alternative titles
   */
  async searchMovies(params: SearchMovieParams): Promise<MoviesResponse> {
    const defaultParams = {
      include_adult: false,
      page: 1,
    };

    const finalParams = { ...defaultParams, ...params };
    return this.makeRequest<MoviesResponse>('/search/movie', finalParams);
  }

  /**
   * Get movies by genre (convenience method)
   */
  async getMoviesByGenre(genreId: number, page: number = 1): Promise<MoviesResponse> {
    return this.discoverMovies({ 
      with_genres: genreId.toString(),
      page 
    });
  }

  /**
   * Get movies by year (convenience method)
   */
  async getMoviesByYear(year: number, page: number = 1): Promise<MoviesResponse> {
    return this.discoverMovies({
      year,
      page
    });
  }

  /**
   * Get highly rated movies with minimum vote count (convenience method)
   */
  async getHighlyRatedMovies(page: number = 1, minVoteCount: number = 1000): Promise<MoviesResponse> {
    return this.discoverMovies({ 
      sort_by: 'vote_average.desc',
      'vote_count.gte': minVoteCount,
      page 
    });
  }

  /**
   * Get image URL for posters and backdrops
   */
  getImageUrl(path: string | null, size: 'w200' | 'w500' | 'w780' | 'original' = 'w500'): string | null {
    if (!path) return null;
    return `https://image.tmdb.org/t/p/${size}${path}`;
  }

  /**
   * Format release date to Brazilian format
   */
  formatBrazilianDate(dateString: string): string {
    if (!dateString) return 'Data não disponível';
    
    try {
      const date = new Date(dateString);
      return date.toLocaleDateString('pt-BR');
    } catch (error) {
      return dateString;
    }
  }

  /**
   * Format runtime to hours and minutes in Portuguese
   */
  formatRuntime(minutes: number): string {
    if (!minutes) return 'Duração não disponível';
    
    const hours = Math.floor(minutes / 60);
    const mins = minutes % 60;
    
    if (hours > 0) {
      return `${hours}h ${mins}min`;
    }
    return `${mins}min`;
  }

  /**
   * Format currency to Brazilian Real
   */
  formatCurrency(amount: number): string {
    if (!amount) return 'Não informado';
    
    return new Intl.NumberFormat('pt-BR', {
      style: 'currency',
      currency: 'USD'
    }).format(amount);
  }
}

// Export singleton instance
export const tmdbService = new TMDBService();
export default tmdbService;

File: kmp/composeApp/src/commonMain/kotlin/com/filmcatalog/kmp/data/remote/TMDBService.kt
package com.filmcatalog.kmp.data.remote

import com.filmcatalog.kmp.data.model.Movie
import com.filmcatalog.kmp.data.model.MoviesResponse
import kotlinx.serialization.Serializable
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.logging.*
import io.ktor.client.request.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.Json

@Serializable
data class GenresResponse(
    val genres: List<GenreItem>
)

@Serializable
data class GenreItem(
    val id: Int,
    val name: String
)

class TMDBService {
    private val baseUrl = "https://api.themoviedb.org/3/"
    private val apiToken = "eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJiZmUwMWE1Y2ZmYzBmNTY3MDZhYmM1NmU4YzNlNzA4ZCIsIm5iZiI6MTc1ODMzNDY5Mi4xNjQsInN1YiI6IjY4Y2UwZWU0MWVjNDQzYjEwMGNkNmU1ZSIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.3Q1ybsIUmSvSikNED11dtMIQ-beSVXfJOMZE-jO8HKg"
    
    private val client = HttpClient {
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
                isLenient = true
            })
        }
        install(Logging) {
            level = LogLevel.INFO
        }
    }
    
    fun getImageUrl(path: String?, size: String = "w500"): String? {
        return if (path != null) {
            "https://image.tmdb.org/t/p/$size$path"
        } else null
    }
    
    suspend fun getPopularMovies(page: Int = 1): MoviesResponse {
        return getMovies("movie/popular", page)
    }
    
    suspend fun getTopRatedMovies(page: Int = 1): MoviesResponse {
        return getMovies("movie/top_rated", page)
    }
    
    suspend fun getNowPlayingMovies(page: Int = 1): MoviesResponse {
        return getMovies("movie/now_playing", page)
    }
    
    suspend fun getUpcomingMovies(page: Int = 1): MoviesResponse {
        return getMovies("movie/upcoming", page)
    }
    
    suspend fun getGenres(): GenresResponse {
        return try {
            val response = client.get("${baseUrl}genre/movie/list") {
                header("Authorization", "Bearer $apiToken")
                header("accept", "application/json")
                parameter("language", "pt-BR")
            }
            
            response.body()
        } catch (e: Exception) {
            GenresResponse(genres = emptyList())
        }
    }
    
    suspend fun discoverMovies(
        page: Int = 1,
        withGenres: String? = null,
        sortBy: String = "popularity.desc"
    ): MoviesResponse {
        return try {
            val response = client.get("${baseUrl}discover/movie") {
                header("Authorization", "Bearer $apiToken")
                header("accept", "application/json")
                parameter("language", "pt-BR")
                parameter("page", page)
                parameter("region", "BR")
                parameter("sort_by", sortBy)
                parameter("include_adult", false)
                if (withGenres != null) {
                    parameter("with_genres", withGenres)
                }
            }
            
            val moviesResponse: MoviesResponse = response.body()
            
            // Add image URLs to movies
            val moviesWithImages = moviesResponse.results.map { movie ->
                movie.copy(
                    posterImageUrl = getImageUrl(movie.posterPath),
                    backdropImageUrl = getImageUrl(movie.backdropPath, "w780")
                )
            }
            
            moviesResponse.copy(results = moviesWithImages)
        } catch (e: Exception) {
            MoviesResponse(
                page = 1,
                results = emptyList(),
                totalPages = 0,
                totalResults = 0
            )
        }
    }
    
    suspend fun searchMovies(query: String, page: Int = 1): MoviesResponse {
        return try {
            val response = client.get("${baseUrl}search/movie") {
                header("Authorization", "Bearer $apiToken")
                header("accept", "application/json")
                parameter("language", "pt-BR")
                parameter("page", page)
                parameter("query", query)
                parameter("include_adult", false)
            }
            
            val moviesResponse: MoviesResponse = response.body()
            
            // Add image URLs to movies
            val moviesWithImages = moviesResponse.results.map { movie ->
                movie.copy(
                    posterImageUrl = getImageUrl(movie.posterPath),
                    backdropImageUrl = getImageUrl(movie.backdropPath, "w780")
                )
            }
            
            moviesResponse.copy(results = moviesWithImages)
        } catch (e: Exception) {
            MoviesResponse(
                page = 1,
                results = emptyList(),
                totalPages = 0,
                totalResults = 0
            )
        }
    }
    
    private suspend fun getMovies(endpoint: String, page: Int): MoviesResponse {
        return try {
            val response = client.get("$baseUrl$endpoint") {
                header("Authorization", "Bearer $apiToken")
                header("accept", "application/json")
                parameter("language", "pt-BR")
                parameter("page", page)
                parameter("region", "BR")
            }
            
            val moviesResponse: MoviesResponse = response.body()
            
            // Add image URLs to movies
            val moviesWithImages = moviesResponse.results.map { movie ->
                movie.copy(
                    posterImageUrl = getImageUrl(movie.posterPath),
                    backdropImageUrl = getImageUrl(movie.backdropPath, "w780")
                )
            }
            
            moviesResponse.copy(results = moviesWithImages)
        } catch (e: Exception) {
            MoviesResponse(
                page = 1,
                results = emptyList(),
                totalPages = 0,
                totalResults = 0
            )
        }
    }
}